ipfx/feature_vectors.py

```diff
@@ -97,21 +97,53 @@ def extract_feature_vectors(data_set,
     return all_features
 
 
-def extract_multipatch_feature_vectors(lsq_sweeps, lsq_start, lsq_end,
+def extract_multipatch_feature_vectors(lsq_supra_sweeps, lsq_supra_start, lsq_supra_end,
+                                       lsq_sub_sweeps, lsq_sub_start, lsq_sub_end,
                                        target_sampling_rate=50000, ap_window_length=0.003):
-    lsq_spx, lsq_spfx = dsf.extractors_for_sweeps(lsq_sweeps, start=lsq_start, end=lsq_end)
-    lsq_an = spa.LongSquareAnalysis(lsq_spx, lsq_spfx, subthresh_min_amp=-100)
-    lsq_features = lsq_an.analyze(lsq_sweeps)
-
-    all_features = feature_vectors(lsq_sweeps, None, None,
-                                   lsq_features, None, None,
-                                   lsq_start, lsq_end,
-                                   lsq_spx,
-                                   target_sampling_rate=target_sampling_rate,
-                                   ap_window_length=ap_window_length)
+    lsq_supra_spx, lsq_supra_spfx = dsf.extractors_for_sweeps(lsq_supra_sweeps, start=lsq_supra_start, end=lsq_supra_end)
+    lsq_supra_an = spa.LongSquareAnalysis(lsq_supra_spx, lsq_supra_spfx, subthresh_min_amp=-100., require_subthreshold=False)
+    lsq_supra_features = lsq_supra_an.analyze(lsq_supra_sweeps)
+
+    lsq_sub_spx, lsq_sub_spfx = dsf.extractors_for_sweeps(lsq_sub_sweeps, start=lsq_sub_start, end=lsq_sub_end)
+    lsq_sub_an = spa.LongSquareAnalysis(lsq_sub_spx, lsq_sub_spfx, subthresh_min_amp=-100., require_suprathreshold=False)
+    lsq_sub_features = lsq_sub_an.analyze(lsq_sub_sweeps)
+
+    all_features = feature_vectors_multipatch(
+        lsq_supra_sweeps, lsq_supra_features,
+        lsq_supra_start, lsq_supra_end,
+        lsq_sub_sweeps, lsq_sub_features,
+        lsq_sub_start, lsq_sub_end,
+        lsq_supra_spx,
+        target_sampling_rate=target_sampling_rate,
+        ap_window_length=ap_window_length
+    )
     return all_features
 
 
+def feature_vectors_multipatch(lsq_supra_sweeps, lsq_supra_features,
```

Comment:
This function has a very complex input and does not seem to represent a reusable action.  Writing tests for this function with so many arguments is difficult especially since it accepts a spike extractor object. Thus, I would suggest to eliminate `feature_vectors_multipatch` and place the code inside `extract_multipatch_feature_vector` (the same is for `extract_feature_vector`). Later on I would like us to refactor extract_feature_vector/extract_multipatch_feature_vector to make them more maintainable.

Reply:
I’m not quite sure what you mean about it not being a reusable action - it’d be run for hundreds or thousands of cells, so that’s a repeated set of steps that seems appropriate to encapsulate? Why is a long list of arguments inherently an issue?

But I’m open to refactoring it. If anything, maybe it’s more appropriate to move most of the logic of the `extract_feature_vector()` and `feature_vectors()` functions (and multipatch equivalents) to the runner scripts and just leave the individual feature vector calculation functions in `feature_vectors.py`.

Reply:
We already largely discussed this, and you latest refactoring is looks really good. I just thought I would still response to your questions here for clarity.  
By not being reusable I meant that the function is called in only one place in the code. But, yes you are correct it is certainly reusable because it is being used to process many cells in a loop.
Generally, long list of arguments  makes it harder to understand, maintain and test functions. But, I am not suggesting that you should necessarily strive to reduce the number of arguments in the functions. Trying to make functions doing just "one thing" should naturally reduce number of arguments.